### PR TITLE
added multi-softpatching support (#9947) + show an OSD message on successful softpatch load

### PR DESCRIPTION
**Pull request by eadmaster**
## Description

Fixes https://github.com/libretro/RetroArch/issues/9947

If an ips/bps/ups patch is found, will replace the last char in the filename with a counter and loop until another patch is found (in any format).

Example matching filenames:
````
Contra (Japan).nes
Contra (Japan).ips # applied first
Contra (Japan).ips1 # applied second
Contra (Japan).ups2 # applied third
...
````

EDIT: notice that if one filename breaks the loop, the following patches won't be applied too. e.g.:
````
Contra (Japan).nes
Contra (Japan).ips # applied
Contra (Japan) (Translation).ips1 # partial name match, not applied
Contra (Japan).ups3 # not applied either
...
````

EDIT2: updated to refrect the new extensions format.

### CODE REVIEW

**Comment by keithbowes** [comment]
It needs to be possible to disable the OSD somehow, like a `notification_show_softpatch_load` setting (and an option in the Notification Visibility section of the UI).

**Comment by eadmaster** [comment]
> It needs to be possible to disable the OSD somehow, like a `notification_show_softpatch_load` setting (and an option in the Notification Visibility section of the UI).

do all the OSD messages have an option to disable them?
I don't see one for the savestates and this softpatch notification is not very obtrusive since it shows only once when new content is loaded.

**Comment by eadmaster** [comment]
UPDATE: i still have another  fix to push for this to limit the number of patches to 9 and avoid an out of bounds access, do not merge it yet!

**Comment by eadmaster** [comment]
ok, should be good to merge now...

**Review by Sanaki — CHANGES_REQUESTED** [review]
Just tested this out. The functionality works like a dream, except that the patch number is double incrementing. You can load up to five total right now, in order: ips, ip2, ip4, ip6, ip8. Any odd numbers are skipped over.
```
[INFO] [CONTENT LOAD]: Loading content file: /usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).sfc.
[INFO] Found IPS file in "/usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).ips", attempting to patch ...
[INFO] Found IPS file in "/usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).ip2", attempting to patch ...
[INFO] Found IPS file in "/usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).ip4", attempting to patch ...
[INFO] Found IPS file in "/usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).ip6", attempting to patch ...
[INFO] [CONTENT LOAD]: CRC32: 0x45abffba .
```

File: tasks/task_patch.c
Comment (Sanaki) [inline tasks/task_patch.c:759]: Changing the counter start to 2 and removing the later increment in this loop appears to sort the issue fully.

**Comment by Sanaki** [comment]
Yeah, that's the better solution. Shows what happens when I touch C.

**Comment by eadmaster** [comment]
> Yeah, that's the better solution. Shows what happens when I touch C.

what if i change the following `if( i == 1 )` to `if( i == 2 )` instead ?

**Comment by Sanaki** [comment]
That seems to work fine too.

**Comment by jdgleaver** [comment]
@eadmaster This is very useful functionality! Thanks for working on this :)

There are a few small issues, however:

- This is unsupported when using many toolchains:

```c
   const int OSD_MSG_MAX_LEN = FILENAME_MAX + strlen(msg_hash_to_str(MSG_PATCH_LOADED));
   char osd_msg[OSD_MSG_MAX_LEN]; 
```

The size of `osd_msg` must be a 'real' constant, or this won't compile on several platforms

- I really dislike the practice of modifying function arguments that are meant to be `const` - I would much prefer that you modify copies of the patch file paths

- I think replacing the last character of the patch file extension kinda strips the patch file of its identity. `.ips` is clear - `.ip` with a number could refer to an entirely different file type. I would prefer the index to be appended after the last file extension character.

- In your `for (i = 2; i < 11; i++)` loop, you increment up to `i = 10`. If `i` is 10, then `'0'+i` will not produce a valid character

- Small naming/wording semantics: I think a patch should be 'applied' rather than 'loaded', in the same way that cheats are 'applied'

- We 100% need an option to hide patch notifications. When applying more than a couple, the OSD log spam is unbearable!

I have prepared a diff that fixes all these issues: [soft_patch.zip](https://github.com/libretro/RetroArch/files/6129815/soft_patch.zip)

Could you please apply this on-top of your PR, and then check that you're happy with the changes? If so, I'd then request that you rebase and squash your commits - and then we can go ahead and merge this :)

**Comment by Sanaki** [comment]
Tested the patch out, works beautifully. Definitely appreciate the ability to hide the OSD notifications. Verified it now supports 10 patches (ips, ips1, ips2, and so on). Verified it will skip a non-matching bps in the middle of the chain and continue patching in order after that.

**Comment by jdgleaver** [comment]
@Sanaki Nice! Thank you very much for testing this :)

**Comment by eadmaster** [comment]
> I think replacing the last character of the patch file extension kinda strips the patch file of its identity. .ips is clear - .ip with a number could refer to an entirely different file type. I would prefer the index to be appended after the last file extension character.

[ZSNES](http://www.emulation64.com/view/976/ZSNES-WIP-September-29th-Released/) used these extensions for multipatch support, but i guess the choice was forced by 8+3 DOS filesystem limitations. Still, i think 3 chars filename extensions is more portable...

**Comment by jdgleaver** [comment]
Ah, DOS 8.3 filenames! Yes, I hadn't thought of that...

Hmm... Well maybe, at least until we get the RetroArch DOS port up and running (it's not really usable at present), we can stick with the `.ips1` naming? No currently supported platform will have problems with that :)

If we need to support DOS 8.3, then we should probably also support all the manifold naming schemes listed here: https://github.com/snes9xgit/snes9x/issues/217#issuecomment-325655551 - but that represents quite a significant performance overhead....

**Comment by eadmaster** [comment]
> we can stick with the .ips1 naming? No currently supported platform will have problems with that :)

At this point i prefer the `.XXX.ips` extension used by snes9x, which keeps the regular `.ips` extension.

**Comment by Sanaki** [comment]
If it supports up to 999 patches, at that point the popup would probably need to be modified to just say "Applied 327 patches" or whatever. Otherwise you run the potential of rolling notifications for literal hours.

**Comment by eadmaster** [comment]
update: I've added support for `*.ipX` and `*.XXX.ips` patches filenames with additional loops.

When combined this gives up to `8*3` patches support.

@Sanaki: i've noticed you started the patch index counter at 0, while in my original code it was 2 (`.ips0` and `.ips1` have a confusing ordering for me vs the single `.ips`).

**Comment by Sanaki** [comment]
I merely reported the behavior shown by jdgleaver's patched code. That said, this iteration seems to need a touch more work. Just starting it up with the same four base patches I had before, this was the result:
```
[INFO] [CONTENT LOAD]: Loading content file: /usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).sfc.
[INFO] Found IPS file in "/usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).ips", attempting to patch ...
[INFO] Found IPS file in "/usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).ips3", attempting to patch ...
/usr/local/games/ROMs/MSU-1/Final Fantasy V/Final Fantasy V (Japan).ip2.001.ips
[INFO] [CONTENT LOAD]: CRC32: 0xd3824c42 .
```
Regardless of whether that naming scheme is still correct, that behavior is not.

**Comment by eadmaster** [comment]
> I merely reported the behavior shown by jdgleaver's patched code. 

Sorry, i've made confusion with the original author of the patch :-)

> That said, this iteration seems to need a touch more work. Just starting it up with the same four base patches I had before, this was the result:


@Sanaki: shoud be fixed now...

**Comment by Sanaki** [comment]
Works now (ips, ips2, ips3, ips4), but I still get the odd `.ip2.001.ips` line in my output. Not sure what that one's about, or if it has any actual effect.

**Comment by eadmaster** [comment]
> Works now (ips, ips2, ips3, ips4), but I still get the odd `.ip2.001.ips` line in my output. Not sure what that one's about, or if it has any actual effect.

yeah, sorry... forgot to restore the `ips` extension before the final loop...

**Comment by eadmaster** [comment]
Now i'm thinking it should be better to replace all the direct string manipulations with `snprintf()` calls for better readability.

If you have other improvements or fixes you should be able to send PRs [here](https://github.com/eadmaster/RetroArch).

**Comment by twinaphex** [comment]
Merge conflict.

Also, I'd measure performance first of direct string manipulation vs. snprintf. If it's just a few trivial sections where we use direct string manipulation, that is probably worth it for the better performance.

**Comment by jdgleaver** [comment]
I don't like the way this PR is going...

With my diff applied, this was a simple solution that was ready to merge. Now the performance overheads are escalating, and `*.XXX.ips` format support is half-baked, since users will expect `*.999.ips` to work, and it doesn't. Moreover, there is a confusing inconsistency in the indexing - `*.XXX.ips` starts from 1, while the other formats start from 2.

If you want to support multiple formats like this, then serious thought needs to go into the efficiency of the whole process. `path_is_valid()` is slow on many platforms - and now we're calling it at least 9 times after the first patch is found. Nor is this cleanly extensible if we want to support additional naming schemes (and as soon as you bring snprintf() into the mix, it gets even worse). And as Sanaki points out, if we're supporting arbitrary files then there must be only one notification with the total number of patches applied.

I guess my point is that we either have a simple extension to the existing code (which is what was originally proposed), or the whole thing needs to be rewritten properly, and rigorously, with a new library for finding/applying patches and options for users to disable any functionality that unduly increases performance overheads.

**Comment by eadmaster** [comment]
>  ... *.XXX.ips format support is half-baked, since users will expect *.999.ips to work, and it doesn't. Moreover, there is a confusing inconsistency in the indexing - *.XXX.ips starts from 1, while the other formats start from 2.

This can easily be fixed.

> path_is_valid() is slow on many platforms - and now we're calling it at least 9 times after the first patch is found. 

Why 9 times? It will break immediately after a patch is not found, so only 3 times if there are no other patches...

> Nor is this cleanly extensible if we want to support additional naming schemes (and as soon as you bring snprintf() into the mix, it gets even worse).

Actually `snprintf()` shoud make the code more readable with less direct string manipulations, and so should be easier to add more naming schemes later.

> And as Sanaki points out, if we're supporting arbitrary files then there must be only one notification with the total number of patches applied.

Ok, this definitively makes sense.

Btw, i propose these changes:
 - comment out *.XXX.ips format
-  keep direct string manipulations (no `snprintf()`)

**Comment by Sanaki** [comment]
Far as I can tell, this is fine as-is, though it still has a conflict in need of resolution before it's able to be merged.

**Comment by jdgleaver** [comment]
Upon further reflection, I think that indexed patch files are somewhat counter-productive, in that they create a maintenance burden for both the developer and the user - i.e. devs have to consider all possible indexing formats, and users have to rename their patches to meet specific requirements (such that the functionality of each individual patch is unclear). I think a better long term solution would be as follows:

- On load content, check for the existence of a directory matching the content file name
- If directory exists, load all patch files inside it (in alphabetical order)
- If it does not, fall back to the existing behaviour of loading `<content_name>.<ips/bps/etc>`
- Add a menu option to disable patching globally, so users who don't need this functionality are unencumbered by the additional performance overheads

This would allow users to simply dump patches from romhacking.net into a directory, with each patch maintaining it's proper name - such that it is easy to add, remove and update them. Moreover, all the duplicate loops and checks required for different indexing formats are replaced by a single directory existence check + a directory listing (one standard piece of future-proof code). If users require patches to be applied in a particular order, they can just prefix the 'correct' patch filename with a number.

**Comment by eadmaster** [comment]
>     * Add a menu option to disable patching globally, so users who don't need this functionality are unencumbered by the additional performance overheads
> 

As alternative i was thinking of a compile-time define, so the whole multi-patch feature could be disabled on slower target machines. But again, to avoid slowing down this PR further let's just save it for future iterations.

Btw creating a subdir to only load 2 patches sounds a burden for me as well, i still prefer partial filename matching. 

As an user i'm having a similar issue with savestates that you cannot rename to store some description (e.g. `"Zelda (U) [before the final boss].state9"`) or they won't load anymore, so partial filename matching could be reused in this case as well.

>though it still has a conflict in need of resolution before it's able to be merged.

I'll fix it and squash the commits asaic.

**Comment by Sanaki** [comment]
>     * If directory exists, load all patch files inside it (in alphabetical order)

This seems like it would interfere with the future implementation of a highly demanded feature: the ability to manually select from multiple patches to apply. We hear requests for this in discord multiple times a week. As a common example, Super Mario World has literally hundreds of different hacks that any given player may want to pick from. While I realize we're not there yet, having conflicting functionality could lock that prospect off entirely, or cause significant woes when changing the implemented scheme.

I agree that it would be preferable to be able to name the patches somehow, the original issue I opened purely specified this scheme because it was already a standard implemented in Snes9x. Perhaps a reasonable compromise for the time being would be to allow `<rom name>+<arbitrary name>.ips#`. The choice of + as the delimiter is somewhat arbitrary there. So long as it's not a space, hyphen, underscore, or parenthesis/bracket, just about anything else should suffice.

**Comment by jdgleaver** [comment]
> If directory exists, load all patch files inside it (in alphabetical order)

I don't see how this interferers with selecting patches via the menu. In fact, it makes things a whole lot easier and more performant.

By default, all patches would be loaded. If the user enables 'patch selection', then the menu would allow items inside the `content_name` directory to be toggled.

If we don't have a `content_name` subdirectory, then we have to scan the entire rom directory every time and filter out all those `<rom name>+<patch name>.ips#` files. This is slow as hell, and not really something that I would accept in the code base (it's bad enough that we currently do something similar with save states, but I hope to change this if I ever get time...)

**Comment by Sanaki** [comment]
If that's how you see that potentially implemented, that could work. I assume the prior method of just slapping a patch next to the rom without the subdirectory would continue to work, yes? Otherwise I'd have to add hundreds of exceptions to my rom manager, and I'd rather avoid that, as I assume many others would.

**Comment by jdgleaver** [comment]
> I assume the prior method of just slapping a patch next to the rom without the subdirectory would continue to work, yes?

Absolutely 100% yes! As I said:

> - If directory exists, load all patch files inside it (in alphabetical order)
> - If it does not, fall back to the existing behaviour of loading <content_name>.<ips/bps/etc>

We never remove existing functionality unless there are overwhelming reasons to do so. In this case, patch subdirectories would be entirely optional - and all old patch setups would continue to work without intervention.

**Comment by eadmaster** [comment]
UPDATE: conflict solved but now i'm having problems with the squash, see if you can help me...

2 things to consider for future reference:
 1. multipatch support may lead to incompatible patches combinations and/or unwanted behaviors (e.g. crashes)
 2. soft-patchin also has the side-effect that RA will produce sram saves and savestates named after the base ROM. This again could produce compatibility issues when the same saves are loaded on an unpatched (or differently-patched) base ROM (there is no way to ensure saves-base ROM compatiblity. Mednafen is appending the ROM hash to the save filenames to prevent this).

Because of 2. i've stopped using soft-patching completely and reverted to hard-patched ROMs named with the `[T+Eng]` tag. So i don't think in the future i'll continue improving this feature but i'll support [this playlist approach](https://github.com/libretro/RetroArch/issues/10135) instead.

**Comment by Sanaki** [comment]
Build error on x86_64 Linux:
```
CC tasks/task_patch.c
distcc[12447] ERROR: compile tasks/task_patch.c on localhost failed
tasks/task_patch.c: In function ‘patch_content’:
tasks/task_patch.c:839:45: error: lvalue required as left operand of assignment
  839 |          name_ips_indexed[name_ips_len] - 1 = index_char;
      |                                             ^
tasks/task_patch.c:840:45: error: lvalue required as left operand of assignment
  840 |          name_bps_indexed[name_bps_len] - 1 = index_char;
      |                                             ^
tasks/task_patch.c:841:45: error: lvalue required as left operand of assignment
  841 |          name_ups_indexed[name_ups_len] - 1 = index_char;
      |                                             ^
Makefile:204: recipe for target 'obj-unix/release/tasks/task_patch.o' failed
make: *** [obj-unix/release/tasks/task_patch.o] Error 1
```

**Comment by eadmaster** [comment]
the more i use git the more i am hating it...

**Comment by jdgleaver** [comment]
I'm struggling with a PR of my own right now, but I'll check this over properly as soon as I'm done. I can see a couple of very trivial things right away, though, so I'll leave some comments in the code.

> @jdgleaver: maybe using a static buffer for the filenames is faster than using malloc?

It would be faster, but we have to be careful about allocating lots of large static arrays. These are problematic on platforms with limited stack space... (i.e. it can sometimes break the console ports entirely)

File: config.def.h
Comment (jdgleaver) [inline config.def.h:847]: This is defined twice!

File: tasks/task_patch.c
Comment (jdgleaver) [inline tasks/task_patch.c:786]: Why are these `len + 10`?

File: config.def.h
Comment (eadmaster) [reply config.def.h]: fixed

File: tasks/task_patch.c
Comment (eadmaster) [reply tasks/task_patch.c]: fixed and resquashed

**Comment by jdgleaver** [comment]
@eadmaster The rebase has failed again....

**Comment by eadmaster** [comment]
> @eadmaster The rebase has failed again....

Since this PR is touching various files currently it is likely to happen again and again. I think we should rendezvous at a specific time and date for the next rebase+merge.

**Comment by jdgleaver** [comment]
@eadmaster That's not how it works...

These problems are only occurring because you are not working in a feature branch. I already told you how to do this: https://github.com/libretro/gambatte-libretro/pull/167#issuecomment-735218802 (start reading from `For future reference, it's good practice to work in a feature branch....`)

I'm not sure if the PR as it stands is even fixable now...

**Comment by eadmaster** [comment]
Do you think it may help if i move the new code into a branch and send another PR? (This one will be closed obviusly)

**Comment by jdgleaver** [comment]
Yes, that would be great - it should solve these merge issues quite effectively :)

**Comment by eadmaster** [comment]
New PR sent.

**Comment by i30817** [comment]
There are roms that have filenames with '.' before the extension. Will this break on those and consider the second to last '.' a 'patch marker'?

**Comment by eadmaster** [comment]
> There are roms that have filenames with '.' before the extension. Will this break on those and consider the second to last '.' a 'patch marker'?

It won't break on those filenames with a `"."` inside, the extension is 


> I'm also not surprised by the problems making you go back to hardpatches. ...

After some experiments, i resorted to have multiple patched copies of the same base ROM vs a single base ROM + multiple patches:
 Less files = less cluttering to manage.

In the end, i think it still doesn't hurt having multiple options to choose. Maybe soft-patching is a bit easier
to use than hard-patching, since you do not need a tool to actually apply the patch.

As for the checksums, there is a [method to read the crc32 of the loaded content](https://github.com/libretro/RetroArch/blob/master/tasks/task_content.c#L2690) which could be used to have a stronger association with the saves. I still think the [Mednafen solution is pretty solid and easy to implement](https://mednafen.github.io/documentation/#filesys.fname_sav), since it does not require any db, but should be optional/keep the current filename scheme as default. Btw, this should be moved to a separate issue...

**Comment by i30817** [comment]
The reason CRC is not used is that it's not amortized all the time (ie: outside zips) and even inside zips, RA has no clue which singular file is 'uniquely identifying' so it's just as likely to crc the .cue file of a iso, _and_ speaking of that, crc32 large file (not amortized) takes too long.

This would be fine if the method was done only for data which you're sure is amortized correctly (ie: a internal checksum of a chd), but it has failure points otherwise.

**Comment by Sanaki** [comment]
Please continue this conversation in an issue if you want to propose changes. This isn't the place for it.

**Comment by andiandi13** [comment]
@eadmaster It'd be great if patches could be detected as long as it ends with .ips$

That'd be a workaround to name our patches and recognize them.

e.g.

Castlevania - Harmony of Dissonance (USA).gba
Castlevania - Harmony of Dissonance (USA).ips-no-blue_outlines

**Comment by Sanaki** [comment]
While much more relevant to the topic at hand, please open an issue with a reference to this PR if you'd like to discuss further improvements to it. The PR is already merged and bug-free, so conversation about future changes doesn't fit well here and won't be seen by any others who might be interested in the topic.

Also feel free to tag me if you do, I'm always interested in further improvements to the softpatching system.

**Comment by andiandi13** [comment]
@Sanaki I will thank you for the advice

**Pull request closed without merging**
